source/darwin/nw_socket.c

```diff
@@ -1039,82 +1035,121 @@ static void s_process_listener_success_task(struct aws_task *task, void *args, e
     struct nw_socket *listener_nw_socket = task_args->nw_socket;
+    int error = task_args->error_code;
+
+    AWS_FATAL_ASSERT(listener_nw_socket && listener_nw_socket->mode == NWSM_LISTENER);
+
+    AWS_LOGF_TRACE(
+        AWS_LS_IO_SOCKET,
+        "id=%p handle=%p: start to process incoming connection.",
+        (void *)listener_nw_socket,
+        (void *)listener_nw_socket->os_handle.nw_listener);
 
     if (status == AWS_TASK_STATUS_RUN_READY) {
+        s_lock_base_socket(listener_nw_socket);
+        struct aws_socket *listener = listener_nw_socket->base_socket_synced_data.base_socket;
+        AWS_FATAL_ASSERT(listener->accept_result_fn);
+        struct aws_socket *new_socket = NULL;
 
-        if (listener_nw_socket) {
-            s_lock_base_socket(listener_nw_socket);
-            struct aws_socket *listener = listener_nw_socket->base_socket_synced_data.base_socket;
-
-            struct aws_socket *new_socket = aws_mem_calloc(allocator, 1, sizeof(struct aws_socket));
-            struct aws_socket_options options = listener->options;
-            int error = aws_socket_init(new_socket, allocator, &options);
-            if (error) {
-                aws_mem_release(allocator, new_socket);
-                nw_release(task_args->new_connection);
-                if (listener->accept_result_fn) {
-                    listener->accept_result_fn(listener, task_args->error_code, NULL, task_args->user_data);
-                }
-            } else {
-                new_socket->io_handle.data.handle = task_args->new_connection;
+        if (error) {
+            goto incoming_listener_error_cleanup;
+        }
 
-                new_socket->io_handle.set_queue = s_client_set_dispatch_queue;
+        new_socket = aws_mem_calloc(allocator, 1, sizeof(struct aws_socket));
+        struct aws_socket_options options = listener->options;
+        error = aws_socket_init(new_socket, allocator, &options);
+        if (error) {
+            goto incoming_listener_error_cleanup;
+        }
 
-                nw_endpoint_t endpoint = nw_connection_copy_endpoint(task_args->new_connection);
-                const char *hostname = nw_endpoint_get_hostname(endpoint);
-                uint16_t port = nw_endpoint_get_port(endpoint);
+        nw_endpoint_t endpoint = nw_connection_copy_endpoint(task_args->new_connection);
+        const char *hostname = nw_endpoint_get_hostname(endpoint);
+        uint16_t port = nw_endpoint_get_port(endpoint);
 
-                if (hostname != NULL) {
-                    size_t hostname_len = strlen(hostname);
-                    size_t buffer_size = AWS_ARRAY_SIZE(new_socket->remote_endpoint.address);
-                    size_t to_copy = aws_min_size(hostname_len, buffer_size);
-                    memcpy(new_socket->remote_endpoint.address, hostname, to_copy);
-                    new_socket->remote_endpoint.port = port;
-                }
+        if (hostname != NULL) {
+            size_t address_strlen;
```

Comment:
We almost never want to use libc memory/string functions directly.  Let's just use aws_byte_cursor_from_c_str and make a local aws_byte_buf that refers to the memory and use an appropriate copy function.

Reply:
Not really sure what is this mean...? As the dest address is a char array, I don't see how using a local aws_byte_buf would help. Am I missing something?

Reply:
Updated to use aws_byte_buf, while I'm not sure if that is necessary... 

```            
            struct aws_byte_buf hostname_buf = aws_byte_buf_from_c_str(hostname);
            struct aws_byte_buf address_buf = aws_byte_buf_from_empty_array(new_socket->remote_endpoint.address, AWS_ADDRESS_MAX_LEN);
            aws_byte_buf_write_from_whole_buffer(&address_buf, hostname_buf);
            aws_byte_buf_clean_up(&address_buf);
            aws_byte_buf_clean_up(&hostname_buf);
```